staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go

```diff
@@ -144,3 +144,78 @@ func (t *prefixTransformers) TransformToStorage(data []byte, context Context) ([
 	prefixedData = append(prefixedData, result...)
 	return prefixedData, nil
 }
+
+// LocationTransformer holds a transformer interface and the location of the resources it has to be used for.
+type LocationTransformer struct {
+	Location    []byte
+	Transformer Transformer
+}
+
+type locationTransformers struct {
+	transformers []LocationTransformer
+	err          error
+}
+
+var _ Transformer = &locationTransformers{}
+
+// NewLocationTransformers supports the Transformer interface by first filtering transformers on
+// whether they are configured to write to the given location. This is done by detecting which transformer's
+// location is a prefix of this resource's location, and choosing the most specific of those locations.
+func NewLocationTransformers(err error, transformers ...LocationTransformer) Transformer {
+	if err == nil {
+		err = fmt.Errorf("the provided value not match any of the supported transformers")
+	}
+	return &locationTransformers{
+		transformers: transformers,
+		err:          err,
+	}
+}
+
+// TransformFromStorage collects all possible transformers (based on their location parameter) and
+// attempts to decrypt using each of them
+func (t *locationTransformers) TransformFromStorage(data []byte, context Context) ([]byte, bool, error) {
+	for _, transformerID := range longestMatchingTransformers(&t.transformers, context) {
+		result, stale, err := t.transformers[transformerID].Transformer.TransformFromStorage(data, context)
+		if err == nil {
+			return result, stale, err
+		}
+	}
```

Comment:
There is only one location transformer in the APIserver. It contains all the other providers mentioned in the configuration file. Consider this case:

1. KMS: `/v1/Secrets`
2. AEAD: `/v1/Secrets             # kept for backward compatibility for reading old secrets`

If there is a request for `TransformFromStorage(data, '/v1/Secrets/aSecret')`, `longestMatchingTransformers` shall return both the 1st and the 2nd transformers. Now the secret might have been encrypted with either one of those (the case where no complete migration of data has been done), and thus both of them will be tried one by one. If the secret did not get decrypted by KMS, only in that case shall the AEAD provider be used.